File: ios/TimelessApp/TimelessApp/Core/Network/APIClient.swift
import Foundation

/// HTTP client for API requests with automatic authentication
actor APIClient {
    static let shared = APIClient()

    private let tokenManager = TokenManager.shared
    private let session: URLSession
    private let baseURL: String
    private let decoder: JSONDecoder

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = AppConfig.networkTimeout
        config.timeoutIntervalForResource = AppConfig.networkTimeout * 2

        self.session = URLSession(configuration: config)
        self.baseURL = AppConfig.apiURL
        self.decoder = JSONDecoder()
        self.decoder.keyDecodingStrategy = .convertFromSnakeCase
    }

    // MARK: - Public API

    /// Perform a GET request
    func get<T: Decodable>(_ path: String, authenticated: Bool = true) async throws -> T {
        let request = try await buildRequest(path: path, method: "GET", authenticated: authenticated)
        return try await perform(request)
    }

    /// Perform a POST request with body
    func post<T: Decodable, B: Encodable>(_ path: String, body: B, authenticated: Bool = true) async throws -> T {
        var request = try await buildRequest(path: path, method: "POST", authenticated: authenticated)
        request.httpBody = try JSONEncoder().encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await perform(request)
    }

    /// Perform a POST request without body
    func post<T: Decodable>(_ path: String, authenticated: Bool = true) async throws -> T {
        let request = try await buildRequest(path: path, method: "POST", authenticated: authenticated)
        return try await perform(request)
    }

    /// Perform a PUT request
    func put<T: Decodable, B: Encodable>(_ path: String, body: B, authenticated: Bool = true) async throws -> T {
        var request = try await buildRequest(path: path, method: "PUT", authenticated: authenticated)
        request.httpBody = try JSONEncoder().encode(body)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        return try await perform(request)
    }

    /// Perform a DELETE request
    func delete<T: Decodable>(_ path: String, authenticated: Bool = true) async throws -> T {
        let request = try await buildRequest(path: path, method: "DELETE", authenticated: authenticated)
        return try await perform(request)
    }

    /// Perform a DELETE request without response body
    func delete(_ path: String, authenticated: Bool = true) async throws {
        let request = try await buildRequest(path: path, method: "DELETE", authenticated: authenticated)
        let _: EmptyResponse = try await perform(request)
    }

    // MARK: - Request Building

    private func buildRequest(path: String, method: String, authenticated: Bool) async throws -> URLRequest {
        guard let url = URL(string: "\(baseURL)\(path)") else {
            throw APIError.invalidURL
        }

        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        // Add Authorization header if authenticated
        if authenticated {
            let token = try await tokenManager.getAccessToken()
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        return request
    }

    // MARK: - Request Execution

    private func perform<T: Decodable>(_ request: URLRequest) async throws -> T {
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.invalidResponse
        }

        // Handle error responses
        if httpResponse.statusCode >= 400 {
            throw try handleErrorResponse(data: data, statusCode: httpResponse.statusCode)
        }

        // Handle empty response
        if T.self == EmptyResponse.self {
            return EmptyResponse() as! T
        }

        // Decode response
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIError.decodingFailed(error)
        }
    }

    private func handleErrorResponse(data: Data, statusCode: Int) throws -> APIError {
        // Try to decode error response
        if let errorResponse = try? decoder.decode(APIErrorResponse.self, from: data) {
            switch statusCode {
            case 401:
                return .unauthorized(errorResponse.error.message)
            case 403:
                return .forbidden(errorResponse.error.message)
            case 404:
                return .notFound(errorResponse.error.message)
            default:
                return .serverError(statusCode, errorResponse.error.message)
            }
        }

        // Fallback error
        switch statusCode {
        case 401:
            return .unauthorized("Authentication required")
        case 403:
            return .forbidden("Access denied")
        case 404:
            return .notFound("Resource not found")
        default:
            return .serverError(statusCode, "Server error")
        }
    }
}

// MARK: - Supporting Types

enum APIError: LocalizedError {
    case invalidURL
    case invalidResponse
    case unauthorized(String)
    case forbidden(String)
    case notFound(String)
    case serverError(Int, String)
    case decodingFailed(Error)
    case networkError(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL:
            return "Invalid URL"
        case .invalidResponse:
            return "Invalid response from server"
        case .unauthorized(let message):
            return message
        case .forbidden(let message):
            return message
        case .notFound(let message):
            return message
        case .serverError(let code, let message):
            return "Server error (\(code)): \(message)"
        case .decodingFailed(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        }
    }

    var isUnauthorized: Bool {
        if case .unauthorized = self {
            return true
        }
        return false
    }
}

/// API error response format
private struct APIErrorResponse: Decodable {
    let error: APIErrorDetail
}

private struct APIErrorDetail: Decodable {
    let code: String
    let message: String
}

/// Empty response for requests that don't return data
struct EmptyResponse: Decodable {}

File: android/app/src/main/kotlin/com/timeless/app/core/network/ApiClient.kt
package com.timeless.app.core.network

import android.content.Context
import com.timeless.app.auth.TokenManager
import com.timeless.app.core.config.AppConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import org.json.JSONObject
import java.io.BufferedReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * HTTP client for API requests with automatic authentication
 */
class ApiClient(context: Context) {

    private val tokenManager = TokenManager.getInstance(context)
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    // MARK: - Public API

    /**
     * Perform a GET request
     */
    suspend fun <T> get(
        path: String,
        authenticated: Boolean = true,
        deserializer: (String) -> T
    ): Result<T> {
        return request("GET", path, null, authenticated, deserializer)
    }

    /**
     * Perform a POST request with body
     */
    suspend fun <T> post(
        path: String,
        body: String? = null,
        authenticated: Boolean = true,
        deserializer: (String) -> T
    ): Result<T> {
        return request("POST", path, body, authenticated, deserializer)
    }

    /**
     * Perform a PUT request with body
     */
    suspend fun <T> put(
        path: String,
        body: String? = null,
        authenticated: Boolean = true,
        deserializer: (String) -> T
    ): Result<T> {
        return request("PUT", path, body, authenticated, deserializer)
    }

    /**
     * Perform a DELETE request
     */
    suspend fun <T> delete(
        path: String,
        authenticated: Boolean = true,
        deserializer: (String) -> T
    ): Result<T> {
        return request("DELETE", path, null, authenticated, deserializer)
    }

    // MARK: - Request Execution

    private suspend fun <T> request(
        method: String,
        path: String,
        body: String?,
        authenticated: Boolean,
        deserializer: (String) -> T
    ): Result<T> = withContext(Dispatchers.IO) {
        try {
            val url = URL("${AppConfig.apiUrl}$path")
            val connection = url.openConnection() as HttpURLConnection

            connection.requestMethod = method
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("Content-Type", "application/json")
            connection.connectTimeout = (AppConfig.NETWORK_TIMEOUT_SECONDS * 1000).toInt()
            connection.readTimeout = (AppConfig.NETWORK_TIMEOUT_SECONDS * 1000).toInt()

            // Add Authorization header if authenticated
            if (authenticated) {
                val token = tokenManager.getAccessToken()
                if (token != null) {
                    connection.setRequestProperty("Authorization", "Bearer $token")
                } else {
                    return@withContext Result.failure(ApiError.Unauthorized("Not authenticated"))
                }
            }

            // Write body if present
            if (body != null) {
                connection.doOutput = true
                OutputStreamWriter(connection.outputStream).use { writer ->
                    writer.write(body)
                }
            }

            // Read response
            val responseCode = connection.responseCode
            val responseBody = try {
                connection.inputStream.bufferedReader().use(BufferedReader::readText)
            } catch (e: Exception) {
                connection.errorStream?.bufferedReader()?.use(BufferedReader::readText) ?: ""
            }

            // Handle response
            when {
                responseCode in 200..299 -> {
                    Result.success(deserializer(responseBody))
                }
                responseCode == 401 -> {
                    val error = parseErrorResponse(responseBody)
                    Result.failure(ApiError.Unauthorized(error?.message ?: "Authentication required"))
                }
                responseCode == 403 -> {
                    val error = parseErrorResponse(responseBody)
                    Result.failure(ApiError.Forbidden(error?.message ?: "Access denied"))
                }
                responseCode == 404 -> {
                    val error = parseErrorResponse(responseBody)
                    Result.failure(ApiError.NotFound(error?.message ?: "Resource not found"))
                }
                else -> {
                    val error = parseErrorResponse(responseBody)
                    Result.failure(ApiError.ServerError(responseCode, error?.message ?: "Server error"))
                }
            }
        } catch (e: Exception) {
            Result.failure(ApiError.NetworkError(e.message ?: "Network error"))
        }
    }

    private fun parseErrorResponse(body: String): ErrorDetail? {
        return try {
            val json = JSONObject(body)
            val error = json.optJSONObject("error")
            if (error != null) {
                ErrorDetail(
                    code = error.optString("code", "UNKNOWN"),
                    message = error.optString("message", "Unknown error")
                )
            } else {
                null
            }
        } catch (e: Exception) {
            null
        }
    }

    companion object {
        @Volatile
        private var instance: ApiClient? = null

        fun getInstance(context: Context): ApiClient {
            return instance ?: synchronized(this) {
                instance ?: ApiClient(context.applicationContext).also { instance = it }
            }
        }
    }
}

/**
 * API error types
 */
sealed class ApiError(override val message: String) : Exception(message) {
    data class Unauthorized(override val message: String) : ApiError(message)
    data class Forbidden(override val message: String) : ApiError(message)
    data class NotFound(override val message: String) : ApiError(message)
    data class ServerError(val code: Int, override val message: String) : ApiError(message)
    data class NetworkError(override val message: String) : ApiError(message)
    data class DecodingError(override val message: String) : ApiError(message)

    val isUnauthorized: Boolean
        get() = this is Unauthorized
}

/**
 * API error response detail
 */
private data class ErrorDetail(
    val code: String,
    val message: String
)
